python/src/llvm.cc

```diff
@@ -172,12 +175,60 @@ void init_triton_llvm(py::module &&m) {
            });
 
   py::class_<llvm::Function>(m, "function", py::module_local())
+      .def_property_readonly(
+          "name", [](llvm::Function *fn) { return fn->getName().str(); })
       .def("set_calling_conv", &llvm::Function::setCallingConv)
       .def("add_fn_attr", [](llvm::Function *fn, std::string &name,
                              std::string &val) { fn->addFnAttr(name, val); })
-      .def("has_public_visibility",
+
+      .def("is_nvvm_kernel",
```

Comment:
Is there any API in llvm that we could use to determine if it's a nvvm kernel?

Reply:
Checking the visibility to know if the function is a kernel is what we do in the backend. Why can't we just do that?

Reply:
> is what we do in the backend.

Where, and in which backend?

> Why can't we just do that?

This is what I initially tried.  It's possible to do it if you want me to do it this way.  I switched it to this way because if we do it based on visibility, we have to hack the external_libs linker so that everything it links (which is initially public) gets converted to non-public.  I felt like this was cleaner.

> Is there any API in llvm that we could use to determine if it's a nvvm kernel?

Not to my knowledge.

Reply:
> > is what we do in the backend.
> 
> Where, and in which backend?

Here:
https://github.com/openai/triton/blob/main/include/triton/Conversion/TritonGPUToLLVM/Utility.h#L363
It is common to both AMD and NV backends.

> > Why can't we just do that?
> 
> This is what I initially tried. It's possible to do it if you want me to do it this way. I switched it to this way because if we do it based on visibility, we have to hack the external_libs linker so that everything it links (which is initially public) gets converted to non-public. I felt like this was cleaner.
> 

ah interesting, Is this because it is done after linking the external libs? Could we do it before instead?

> > Is there any API in llvm that we could use to determine if it's a nvvm kernel?
> 
> Not to my knowledge.

yes the concept of kernel doesn't really exist in llvm as far as I know.

Reply:
The LLVM IR does not currently mark the non-kernel functions as internal linkage.  https://gist.github.com/jlebar/5afb101e2776a2b21fd0ec822c16a7aa

The TTGIR does, as you show.  https://gist.github.com/jlebar/354ad8b7eb8be8511b311aef68db245a

I am understanding that you're saying you'd like me to figure out how to propagate the TTGIR public/private attribute to LLVM function linkage and then use that to find the kernels, instead of examining the nvvm properties.  I will try this approach.

Reply:
I've changed the PR; please have a look and let me know if it's what you had in mind.